{% extends "base.html" %}

{% block title %}DUB Portal{% endblock %}

{% block scripts %}
    {{ super() }}
    <script>
        $(document).ready(function () {
            $("#table-terms").DataTable({
                "order": [[2, "desc"]], // sort by correlation
                "scrollX": true
            });
            $("#gsea-enrichment").DataTable({
                "order": [[5, "asc"]] // sort by q-value
            });
            $("#dgea").DataTable({
                "order": [[4, "asc"]] // sort by adj p-value
            });
            $("#dgea-gsea").DataTable({
                "order": [[2, "asc"]], // sort by adj p-value
                "scrollX": true
            });
        })
    </script>
{% endblock %}

{% macro boolbox(value) -%}
    {% if value %}
        <span style="color: green;"><i class="fas fa-check-circle"></i></span>
    {% else %}
        <span style="color: red;"><i class="fas fa-times-circle"></i></span>
    {% endif %}
{%- endmacro %}

{% macro badgebox(name, flag) -%}
    {% if flag %}
        <span class="badge badge-primary">{{ name }}</span>
    {% endif %}
{%- endmacro %}


{% block main %}
    <div class="container" style="margin-top: 2.5em; margin-bottom: 2.5em">
        <div class="row justify-content-center">
            <!-- sidebar, which will move to the top on a small screen -->
            <div class="col-sm-3 d-none d-sm-none d-md-block">
                <nav id="toc" data-toggle="toc" class="sticky-top"></nav>
            </div>
            <!-- main content area -->
            <div class="col-sm-9">
                <h1 style="margin-bottom: 1.5rem">{{ record.hgnc_symbol }} Data Analysis</h1>
                <dl class="row">
                    <dt class="col-4">HGNC Gene Name</dt>
                    <dd class="col-8">{{ record.hgnc_name }}</dd>
                    <dt class="col-4">HGNC Gene Symbol</dt>
                    <dd class="col-8">{{ record.hgnc_symbol }}</dd>
                    <dt class="col-4">Identifiers</dt>
                    <dd class="col-8">
                        <a class="badge badge-info"
                           href="https://bioregistry.io/hgnc:{{ record.hgnc_id }}">hgnc:{{ record.hgnc_id }}</a>
                        <a class="badge badge-info"
                           href="https://bioregistry.io/ncbigene:{{ record.xrefs.entrez_id }}">NCBIGene:{{ record.xrefs.entrez_id }}</a>
                        <a class="badge badge-info"
                           href="https://bioregistry.io/uniprot:{{ record.xrefs.uniprot_id }}">uniprot:{{ record.xrefs.uniprot_id }}</a>
                    </dd>
                    {% if 0 < (record.orthologs | length) %}
                        <dt class="col-4">Orthologs</dt>
                        <dd class="col-8">
                            {% if record.orthologs.mgi %}
                                <a class="badge badge-info" data-toggle="tooltip" data-html="true"
                                   title="Mouse Genome Informatics (MGI) is a model organism database listing mouse genes that are homologous to human genes."

                                   href="https://bioregistry.io/mgi:{{ record.orthologs.mgi }}">mgi:{{ record.orthologs.mgi }}
                                </a>
                            {% endif %}
                            {% if record.orthologs.rgd %}
                                <a class="badge badge-info" data-toggle="tooltip" data-html="true"
                                   title="Rat Genome Database (RGD) is a model organism database listing rat genes that are homologous to human genes."
                                   href="https://bioregistry.io/rgd:{{ record.orthologs.rgd }}">rgd:{{ record.orthologs.rgd }}</a>
                            {% endif %}
                        </dd>
                    {% endif %}
                    <dt class="col-4">INDRA Statements</dt>
                    <dd class="col-8">
                        <a class="badge badge-info"
                           href="https://db.indra.bio/statements/from_agents?format=html&subject={{ record.hgnc_symbol }}&type=Deubiquitination">
                            deubiquitinations</a>
                        <a class="badge badge-info"
                           href="https://db.indra.bio/statements/from_agents?format=html&agent={{ record.hgnc_symbol }}">
                            all statements</a>
                    </dd>
                    <dt class="col-4">Pathway Commons</dt>
                    <dd class="col-8">
                        <a class="badge badge-info"
                           href="https://apps.pathwaycommons.org/interactions?source={{ record.hgnc_symbol }}">
                            Search for {{ record.hgnc_symbol }}</a>
                    </dd>
                    <dt class="col-4">Number of Papers</dt>
                    <dd class="col-8">{{ record.papers }} retrieved on {{ date }}</dd>
                </dl>
                <h2>DepMap Analysis</h2>
                <p>
                    The <a href="https://depmap.org/portal/">Dependency Map (DepMap)</a> is a genome-wide pooled
                    CRISPR-Cas9 knockout proliferation screen conducted in more than 700 cancer cell lines spanning many
                    different tumor lineages. Each cell line in the DepMap contains a unique barcode, and each gene
                    knockout is assigned a “dependency score” on a per cell-line basis which quantifies the rate of
                    CRISPR-Cas9 guide drop. It has been found that proteins with similar DepMap scores across cell
                    lines, a phenomenon known as co-dependent genes, have closely related biological functions. This can
                    include activity in the same or parallel pathways or membership in the same protein complex or the
                    same pathway.
                </p>
                <p>
                    We identified the strongest seven co-dependent genes (“Symbol”) for DUBs and ran GO enrichment
                    analysis. We used Biogrid, IntAct, and Pathway Commons PPIDs, and the NURSA protein-protein
                    interaction databases (PPIDs) to determine whether co-dependent genes interact with one another. The
                    “Evidence” column contains the PPIDs in which the interaction appears as well as whether there is
                    support for the association by an INDRA statement. As another approach to identify potential
                    interactors, we looked at proteomics data from the Broad Institute's <a
                        href="https://portals.broadinstitute.org/ccle">Cancer Cell Line Encyclopedia (CCLE)</a> for
                    proteins whose expression across ~375 cell lines strongly correlated with the abundance of each DUB;
                    it has previously been observed that proteins in the same complex are frequently significantly
                    co-expressed. The correlations and associated p-values in the CCLE proteomics dataset are provided.
                    And, we determined whether co-dependent genes yield similar transcriptomic signatures
                    in the Broad Institute's <a href="https://www.broadinstitute.org/connectivity-map-cmap">Connectivity
                    Map (CMap)</a>. A CMap score greater than 90 is considered significantly similar.
                </p>
                <h3>DepMap Correlations</h3>
                <table class="table table-striped table-hover" id="table-terms">
                    <thead>
                    <tr>
                        <th>Symbol</th>
                        <th>Name</th>
                        <th style="text-align: right">DepMap Correlation</th>
                        <th>Evidence</th>
                        <th style="text-align: right">CCLE Correlation</th>
                        <th style="text-align: right">CCLE Z-score</th>
                        <th style="text-align: right">CCLE <i>p</i>-value (adj)</th>
                        <th style="text-align: right">CCLE Significant</th>
                        <th style="text-align: right">CMAP Score</th>
                        <th>CMAP Type</th>
                    </tr>
                    </thead>
                    <tbody>
                    {% for row in record.depmap.genes %}
                        <tr>
                            <td>
                                <a data-toggle="tooltip" data-html="true"
                                   title="HGNC:{{ row.hgnc_id }}<br>Click for external information from HGNC."
                                   href="https://bioregistry.io/hgnc:{{ row.hgnc_id }}">
                                    {{ row.hgnc_symbol }}
                                </a>
                            </td>
                            <td>
                                <a data-toggle="tooltip" data-html="true"
                                   title="HGNC:{{ row.hgnc_id }}<br>Click for external information from HGNC."
                                   href="https://bioregistry.io/hgnc:{{ row.hgnc_id }}">
                                    {{ row.hgnc_name }}
                                </a>
                            </td>
                            <td style="text-align: right">{{ row.correlation }}</td>
                            <td>
                                {{ badgebox('BioGRID', row.interactions.biogrid) }}
                                {{ badgebox('IntAct', row.interactions.intact) }}
                                {{ badgebox('NURSA', row.interactions.nursa) }}
                                {{ badgebox('DGE', row.interactions.dge) }}
                                {% if row.interactions.pc %}
                                    <a class="badge badge-primary"
                                       href="https://apps.pathwaycommons.org/search?type=Pathway&q={{ record.hgnc_symbol }},{{ row.hgnc_symbol }}">
                                        Pathway Commons <i class="fas fa-link"></i>
                                    </a>
                                {% endif %}
                                {% if 0 < row.interactions.indra %}
                                    <a class="badge badge-primary"
                                       href="https://db.indra.bio/statements/from_agents?format=html&agent0={{ record.hgnc_symbol }}&agent1={{ row.hgnc_symbol }}">
                                        INDRA ({{ row.interactions.indra }}) <i class="fas fa-link"></i>
                                    </a>
                                {% endif %}
                                {% if 0 < row.interactions.reactome %}
                                    <span class="badge badge-primary">
                                        Reactome ({{ row.interactions.reactome }})
                                    </span>
                                {% endif %}
                            </td>
                            {% if row.ccle %}
                                <td style="text-align: right">{{ "%.2f" | format(row.ccle.correlation) }}</td>
                                <td style="text-align: right">{{ "%.2f" | format(row.ccle.z) }}</td>
                                <td style="text-align: right">{{ "%.2e" | format(row.ccle.p_adj) }}</td>
                                <td style="text-align: right">
                                    {% if row.ccle.significant %}
                                        <span style="color: green;"><i class="fas fa-check-circle"></i></span>
                                    {% else %}
                                        <span style="color: red;"><i class="fas fa-times-circle"></i></span>
                                    {% endif %}
                                </td>
                            {% else %}
                                <td></td>
                                <td></td>
                                <td></td>
                                <td></td>
                            {% endif %}
                            {% if row.cmap %}
                                <td style="text-align: right">{{ "%.2f" | format(row.cmap.score) }}</td>
                                <td style="text-align: right">{{ row.cmap.type }}</td>
                            {% else %}
                                <td></td>
                                <td></td>
                            {% endif %}
                        </tr>
                    {% endfor %}
                    </tbody>
                </table>
                <h3>Dependency GO Term Enrichment</h3>
                <p>
                    Gene set enrichment analysis was done on the genes correlated with {{ record.hgnc_symbol }}using the
                    terms from <a href="http://geneontology.org/">Gene Ontology</a> and gene sets derived from the
                    <a href="http://geneontology.org/docs/go-annotations/">Gene Ontology Annotations database</a> via
                    <a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a>.
                </p>
                <p>
                    Using the biological processes and other Gene Ontology terms from well characterized DUBs as a
                    positive control, several gene set enrichment analyses were considered. Threshold-less methods
                    like <a href="https://www.gsea-msigdb.org/gsea/index.jsp">GSEA</a> had relatively poor results.
                    Over-representation analysis with a threshold of of the top 7 highest absolute value Dependency Map
                    correlations yielded the best results and is reported below.
                </p>
                <table class="table table-striped table-hover" id="gsea-enrichment">
                    <thead>
                    <tr>
                        <th>GO Identifier</th>
                        <th>GO Name</th>
                        <th>GO Type</th>
                        <th style="text-align: right"><i>p</i>-value</th>
                        <th style="text-align: right"><i>p</i>-value (adj.)</th>
                        <th style="text-align: right"><i>q</i>-value</th>
                    </tr>
                    </thead>
                    <tbody>
                    {% for row in record.depmap.enrichment %}
                        <tr>
                            <td>
                                <a href="https://bioregistry.io/go:{{ row.identifier }}"
                                   data-toggle="tooltip" data-html="true"
                                   title="GO:{{ row.identifier }}<br>Click for external information from the Gene Ontology.">
                                    GO:{{ row.identifier }}
                                </a>
                            </td>
                            <td>
                                <a data-toggle="tooltip" data-html="true"
                                   title="GO:{{ row.identifier }}<br>Click for external information from the Gene Ontology."
                                   href="https://bioregistry.io/go:{{ row.identifier }}">
                                    {{ row.name }}
                                </a>
                            </td>
                            <td><span class="badge badge-info">{{ row.type }}</span></td>
                            <td style="text-align: right">{{ "%.2e" | format(row.p) }}</td>
                            <td style="text-align: right">{{ "%.2e" | format(row.p_adj) }}</td>
                            <td style="text-align: right">{{ "%.2e" | format(row.q) }}</td>
                        </tr>
                    {% endfor %}
                    </tbody>
                </table>{% if record.knockdown.genes %}
                <h2>Transcriptomics</h2>
                <p>
                    The following table shows the significantly differentially expressed genes after knocking
                    out {{ record.hgnc_symbol }} using CRISPR-Cas9.
                </p>
                <h3>Knockout Differential Expression</h3>
                <table class="table table-striped table-hover" id="dgea">
                    <thead>
                    <tr>
                        <th>Symbol</th>
                        <th>Name</th>
                        <th style="text-align: right">log<sub>2</sub>-fold-change</th>
                        <th style="text-align: right"><i>p</i>-value</th>
                        <th style="text-align: right"><i>p</i>-value (adj.)</th>
                    </tr>
                    </thead>
                    <tbody>
                    {% for row in record.knockdown.genes %}
                        <tr>
                            <td>
                                <a data-toggle="tooltip" data-html="true"
                                   title="HGNC:{{ row.hgnc_id }}<br>Click for external information from HGNC."
                                   href="https://bioregistry.io/hgnc:{{ row.hgnc_id }}">
                                    {{ row.hgnc_symbol }}
                                </a>
                            </td>
                            <td>
                                <a data-toggle="tooltip" data-html="true"
                                   title="HGNC:{{ row.hgnc_id }}<br>Click for external information from HGNC."
                                   href="https://bioregistry.io/hgnc:{{ row.hgnc_id }}">
                                    {{ row.hgnc_name }}
                                </a>
                            </td>
                            <td style="text-align: right">{{ "%.2e" | format(row.log2FoldChange) }}</td>
                            <td style="text-align: right">{{ "%.2e" | format(row.pvalue) }}</td>
                            <td style="text-align: right">{{ "%.2e" | format(row.padj) }}</td>
                        </tr>
                    {% endfor %}
                    </tbody>
                </table>
                <h3>Gene Set Enrichment Analysis</h3>
                {% if record.knockdown.enrichment is none %}
                    <p>
                        There were too few differentially expressed genes to run a meaningful GSEA.
                    </p>
                {% else %}
                    <p>
                        The GSEA method was applied for all genes whose knockout resulted in at least 20 significantly
                        differentially expressed genes.
                    </p>
                    <table class="table table-striped table-hover" id="dgea-gsea">
                        <thead>
                        <tr>
                            <th>ID</th>
                            <th>Name</th>
                            <th style="text-align: right"><i>p</i>-value</th>
                            <th style="text-align: right"><i>p</i>-value (adj.)</th>
                            <th style="text-align: right"><i>log<sub>2</sub></i> Error</th>
                            <th style="text-align: right">ES</th>
                            <th style="text-align: right">NES</th>
                        </tr>
                        </thead>
                        <tbody>
                        {% for row in record.knockdown.enrichment %}
                            <tr>
                                <td>
                                    <a data-toggle="tooltip" data-html="true"
                                       title="{{ row.prefix }}:{{ row.identifier }}<br>Click for external information from {{ row.prefix }}."
                                       href="https://bioregistry.io/{{ row.prefix }}:{{ row.identifier }}">
                                        {{ row.prefix }}:{{ row.identifier }}
                                    </a>
                                </td>
                                <td>
                                    <a data-toggle="tooltip" data-html="true"
                                       title="{{ row.prefix }}:{{ row.identifier }}<br>Click for external information from {{ row.prefix }}."
                                       href="https://bioregistry.io/{{ row.prefix }}:{{ row.identifier }}">
                                        {{ row.name }}
                                    </a>
                                </td>
                                <td style="text-align: right">{{ "%.2e" | format(row.pval) }}</td>
                                <td style="text-align: right">{{ "%.2e" | format(row.padj) }}</td>
                                <td style="text-align: right">{{ "%.2e" | format(row.log2err) }}</td>
                                <td style="text-align: right">{{ "%.2e" | format(row.ES) }}</td>
                                <td style="text-align: right">{{ "%.2e" | format(row.NES) }}</td>
                            </tr>
                        {% endfor %}
                        </tbody>
                    </table>
                {% endif %}
            {% endif %}
                <h2>Literature Mining</h2>
                <p>
                    <a href='https://www.indra.bio'>INDRA</a> was used to automatically assemble known mechanisms
                    related to {{ record.hgnc_symbol }} from literature and knowledge bases.
                    The first section shows only DUB activity and the second shows all other results.
                </p>
                <h3>Deubiquitinase Activity</h3>
                {{ dub_stmt_html | safe }}
                <h3>Other Statements</h3>
                {{ other_stmt_html | safe }}
            </div>
        </div>
    </div>
{% endblock %}
