{% extends "base.html" %}

{% block title %}DUB Portal{% endblock %}

{% block scripts %}
    {{ super() }}
    <script>
        $(document).ready(function () {
            $("#table-terms").DataTable();
        })
    </script>
{% endblock %}

{% block content %}
    <div style="width: 100%" class="text-center">
        <img style="max-height: 220px; margin-top: -50px; margin-bottom: -20px;" src="logo.svg" alt="DUB Portal"/>
    </div>
    <div class="card" style="margin-bottom: 20px;">
        <div class="card-body">
            <a href="/dubportal/about/">
                <img style="max-height: 320px; margin: auto 15px;" class="float-right" src="dubwheel.jpg"
                     alt="A figure showing several modes of information about DUBs."/>
            </a>
            <p>
                Deubiquitinating enzymes (DUBs) are proteases that remove
                <a href="https://en.wikipedia.org/wiki/Ubiquitin">ubiqutin</a> conjugates from proteins, thereby
                regulating protein turnover. However, the majority of substrates and pathways regulated by DUBs remain
                unknown, impeding efforts to prioritize specific proteins for research and drug-development.
            </p>
            <p>
                On this DUB portal, we provide the integration and analysis of publicly available resources as well as
                newly collected transcriptomic data following DUB knockout or inhibition to facilitate the
                exploration of DUB function in oncology.
            </p>
            <p>
                Each DUB is listed with the number of papers in PubMed, the number of INDRA statements about
                deubiquitinating substrates, the number of INDRA statements total (including upstream and downstream
                processes beyond deubiquitination), and the and the fraction of cell lines in the <a
                    href="https://depmap.org/portal/">DepMap</a> that are strongly dependent on each DUB (i.e., cancer
                cell proliferation is strongly impacted by DUB knockout). Blank values in the "Dependent Cell Lines"
                column means the DUB was not knocked out in the DepMap dataset. Additionally, we list the Gene Ontology
                (GO) term most strongly enriched in the top seven co-dependent genes for each DUB and the corresponding
                <i>p</i>-values. Finally, the "DGEA" column contains the number of differentially expressed genes 96
                hours following knockout of the particular DUB in <a
                    href="https://bioregistry.io/EFO:0001209">MDAMB231</a> cells as measured by high-throughput RNAseq.
                Blank values in "DGEA" refer to DUBs that were not knockout out in the RNAseq experiment. Click on an
                individual DUB name for more details and additional data and analysis.
            </p>
        </div>
    </div>
    <table class="table table-striped table-hover" id="table-terms">
        <thead>
        <tr>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the HGNC approved gene symbol for the DUB">
                    DUB
                </span>
            </th>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the HGNC approved gene name for the DUB">
                    DUB Name
                </span>
            </th>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows which DUB family the DUB belongs to annotated by FamPlex">
                    DUB Family
                </span>
            </th>
            <th style="text-align: right">
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the number of papers returned by querying to PubMed with the DUB's gene symbol on {{ date }}">
                    # Papers
                </span>
            </th>
            <th style="text-align: right">
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the number of statements about the DUB's activity retrieved from the INDRA Database on {{ date }}">
                    # DUB Stmts.
                </span>
            </th>
            <th style="text-align: right">
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the number of statements about other aspects of the DUB (besides its activity) retrieved from the INDRA Database on {{ date }}">
                    # Other Stmts.
                </span>
            </th>
            <th style="text-align: right">Dependent Cell Lines</th>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the number of gene sets considered significant by over-representation analysis with multiple hypothesis testing correction using genes depending on a given DUB as the query set.">
                    # Sig. ORA Gene Sets (DepMap)
                </span>
            </th>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows how many genes were significantly differentially expressed due to knockout of the given DUB.">
                DGEA
                </span>
            </th>
            <th>
                <span data-toggle="tooltip" data-html="true"
                      title="Shows the number of gene sets considered significant by GSEA analysis with multiple hypothesis testing correction using genes differentially expressed by knocking out the given DUB as the query set.">
                    # Sig. GSEA Gene Sets (DGEA)
                </span>
            </th>
        </tr>
        </thead>
        <tbody>
        {% for row in rows %}
            <tr>
                <td>
                    <a data-toggle="tooltip" data-html="true"
                       title="HGNC:{{ row.hgnc_id }}<br>Click for more information in the DUB Portal."
                       href="{{ row.hgnc_symbol }}">
                        {{ row.hgnc_symbol }}
                    </a>
                </td>
                <td>
                    <a data-toggle="tooltip" data-html="true"
                       title="HGNC:{{ row.hgnc_id }}<br>Click for more information in the DUB Portal."
                       href="{{ row.hgnc_symbol }}">
                        {{ row.hgnc_name }}
                    </a>
                </td>
                <td>
                    {% if row.dub_family %}
                        <a data-toggle="tooltip" data-html="true"
                           title="FPLX:{{ row.dub_family }}<br>Click for external information from FamPlex."
                           href="https://bioregistry.io/fplx:{{ row.dub_family }}">
                            {{ row.dub_family }}
                        </a>
                    {% endif %}
                </td>
                <td style="text-align: right">
                    <a data-toggle="tooltip" data-html="true"
                       title="Click to see publications for {{ row.hgnc_symbol }} on PubMed."
                       href="https://pubmed.ncbi.nlm.nih.gov/?term={{ row.hgnc_symbol }}[Text Word]">
                        {{ row.papers }}
                    </a>
                </td>
                <td style="text-align: right">{{ row.n_dub_statements }}</td>
                <td style="text-align: right">{{ row.n_other_statements }}</td>
                <td style="text-align: right">
                    {% if row.fraction_cell_lines_dependent == "Unmeasured" %}
                    {% else %}
                        {{ "%.1f"|format(row.fraction_cell_lines_dependent * 100) }}%
                    {% endif %}
                </td>
                <td style="text-align: right">
                    {% if row.enrichment is not none %}
                        {{ row.depmap.enrichment | length }}
                    {% endif %}
                </td>
                <td style="text-align: right">
                    {% if row.knockdown.genes is not none %}
                        {{ row.knockdown.genes | length }}
                    {% endif %}
                </td>
                <td style="text-align: right">
                    {% if row.knockdown.enrichment is not none %}
                        {{ row.knockdown.enrichment | length }}
                    {% endif %}
                </td>
            </tr>
        {% endfor %}
        </tbody>
    </table>
{% endblock %}
