{% extends "base.html" %}

{% block title %}DUB Portal - Data Sources{% endblock %}

{% block content %}
    <h1>Datasets Used</h1>
    <table>
        <tr>
            <td>Resource</td>
            <td>Description</td>
            <td>Key Insights</td>
            <td>Terms / abbreviations</td>
        </tr>
        <tr>
            <td>DGE RNA-seq</td>
            <td>We collected high throughput RNA-seq following knockout or inhibition of individual DUBs. 81 DUB
                knockouts and 7 inhibitors were characterized.
            </td>
            <td>The transcriptomic signatures following loss of individual DUBs.</td>
            <td>
                <p>
                    "3’DGE-seq: 3’ Digital Gene Expression, a type of high-throughput RNA-seq.
                </p>
                <p>DE gene: differentially expressed gene (defined by a false discovery rate (FDR) adjusted p <
                    0.05)."</p>
            </td>
        </tr>

        <tr>
            <td>
                Connectivity Map (CMap)
            </td>
            <td>
                A Broad Institute database of post-perturbation RNA-seq signatures following genetic knockdown or
                over-expression or treatment with inhibitors. The signatures within this resource are comprised of the
                L1000 genes (1000 genes measured or inferred by the assay). Data are available for ~3,000 different
                genes and ~5000 small molecules.
            </td>
            <td>
                Enables identification of genes that, when silenced with RNAi, or overexpressed, have similar
                transcriptional effects as query signatures (e.g., signatures following DUB knockout).
            </td>
            <td>
                <p>
                    "Query mRNA profile: the transcriptomic signature that the user inputs into CMap to retrieve similar
                    signatures.
                </p>
                <p>
                    A tau score quantifies similarity between the query mRNA profile and CMap signatures (tau similarity
                    is computed by counting the number of pairwise mismatches between two ranked lists). CMap recommends
                    a threshold of tau similarity score > 90."
                </p>
            </td>
        </tr>

        <tr>
            <td>
                Dependency Map (DepMap)
            </td>
            <td>
                Broad Institute genome-wide pooled CRISPR-Cas9 knockout screen conducted in more than 700 cancer cell
                lines. Guide RNA dropout is measured as a surrogate for viability.
            </td>
            <td>
                <p>
                    "Enables identification of genes that are essential for cell proliferation as well as which gene
                    knockouts yield similar viability phenotypes across cell lines (co-dependent genes).
                </p>
                <p>
                    The top co-dependent genes can be analyzed to identify protein complexes and pathways enriched in
                    the co-dependent genes."
                </p>
            <td>
                <p>
                    "Genes with similar DepMap scores across cell lines are more likely to have related biological
                    functions, a property known as co-dependency.
                </p>
                <p>
                    Co-dependent genes: gene knockouts that exhibit co-dependency or give similar DepMap scores across
                    cell lines. In our study, we analyzed the top seven co-dependent genes as this gave the largest
                    number of expected results for well-characterized DUBs, but the top five and ten co-dependent genes
                    gave similar results. "
                </p>
            </td>
        </tr>
        <tr>
            <td>
                Cancer Cell Line Encyclopedia (CCLE) Proteomics
            </td>
            <td>
                Previously published baseline proteomics collected in the CCLE cell lines (~375 cell lines). About
                12,000
                proteins were detected.
            </td>
            <td>
                Co-expressed genes based on expression levels across ~375 cell lines. Proteins in the same complex are
                often
                co-expressed to a significant degree across a cell line panel.
            </td>
            <td>
                Co-expressed genes: correlation in abundance across cell lines with FDR < 0.01 and |z-score| > 2.
                Previous
                publication used FDR < 0.01.
            </td>
        </tr>
        <tr>
            <td>BioGRID</td>
            <td>
                Protein-protein interaction database compiling interaction data from multiple sources.
            </td>
            <td>
                Proteins that interact as measured in multiple ways, including affinity capture MS, affinity capture
                Western
                Blotting, and assembly of reconstituted complexes from purified recombinant subunits in vitro.
            </td>
            <td>PPID: protein-protein interaction database.</td>
        </tr>
        <tr>
            <td>IntAct
            </td>
            <td>
                Protein-protein interaction database compiling interaction data from multiple sources.
            </td>
            <td>
                Proteins that interact as measured in multiple ways, including affinity capture MS, affinity capture
                Western
                Blotting, and assembly of reconstituted complexes from purified recombinant subunits in vitro.
            </td>
            <td>
                PPID: protein-protein interaction database.
            </td>
        </tr>
        <tr>
            <td>NURSA</td>
            <td>
                Protein-protein interaction dataset focused on interactions among proteins involved in transcription.
            </td>
            <td>
                Proteins that interact as measured by ~3000 immuno-precipitations (IPs) followed by MS.
            </td>
            <td>"PPID: protein-protein interaction database.</td>
        </tr>
    </table>
{% endblock %}
